### PR TITLE
feat: implement an orbit structure for more efficient geometry computations

### PR DESCRIPTION
**Pull request by imrn99**
# Description

Implement a generic orbit structure that can be used for consistent, standardized cell computation. The algorithm used is detailed in the documentation.

The TwoMap implementation is left unchanged at the moment since replacing its code would probably lead to breaking changes. 

To do:
- [x] add more tests
- [x] add documentation

## Scope

- [x] Code: `honeycomb-core`

## Type of change

- [x] New feature(s)
- [x] Testing

## Other

...

## Necessary follow-up

- [x] Needs documentation
- [x] Other: Benchmark new methods

## Mentions

...

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,26 @@
...
+### New Features
+
+#### honeycomb-core
+
+- `Orbit<'a, const N_MARKS: usize, T: CoordsFloat>` - Generic implementation for
+  2D orbit computations. The structure itself only contains meta-data, the orbit
+  computation is done through the `Iterator` trait implementation.
+- `OrbitPolicy<'a>` - Enum used to specify the beta functions used by an orbit.
+  It currently does not not support compositions.
+- New (temporary?) method for `TwoMap`: `beta_bis`. It works exactly like the
```
Comment (cedricchevalier19) [inline CHANGELOG.md:18]: `beta_bis` name is not very good.
I would rather have `beta_rt` or `beta_runtime` and `beta_ct` (compile time?) to have an clearer idea of whats going on. Or `beta_generic` or `beta_dispatch`.

File: honeycomb-core/src/twomap.rs [hunk 2]
```diff
@@ -102,9 +100,9 @@ const TWO_MAP_BETA: usize = 3;
 /// let (v1, v2, v3): (VertexIdentifier, VertexIdentifier, VertexIdentifier) = (0, 1, 2);
 ///
 /// // place the vertices in space
-/// map.set_vertex(v1, [0.0, 0.0]);
-/// map.set_vertex(v2, [0.0, 10.0]);
-/// map.set_vertex(v3, [10.0, 0.0]);
+/// map.set_vertex(v1, [0.0, 0.0]).unwrap();
+/// map.set_vertex(v2, [0.0, 10.0]).unwrap();
+/// map.set_vertex(v3, [10.0, 0.0]).unwrap();
```
Comment (cedricchevalier19) [inline honeycomb-core/src/twomap.rs:105]: Future work: proper error handling, with `?` for example.

File: honeycomb-core/src/twomap.rs [hunk 3]
```diff
@@ -379,6 +383,32 @@ impl<const N_MARKS: usize, T: CoordsFloat> TwoMap<N_MARKS, T> {
...
+    ///
+    /// The method will panic if *i* is not 0, 1 or 2.
+    ///
+    /// # Example
+    ///
+    /// See [TwoMap] example.
+    ///
+    pub fn beta_bis(&self, i: u8, dart_id: DartIdentifier) -> DartIdentifier {
+        assert!(i < 3);
+        self.betas[dart_id as usize][i as usize]
```
Comment (cedricchevalier19) [inline honeycomb-core/src/twomap.rs:409]: Ok, `betas` is simple enough but a `match` dispatcher that forward to the corresponding compile time beta function could also be considered.

File: honeycomb-core/src/orbits.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+pub enum OrbitPolicy<'a> {
+    /// 0-cell orbit.
+    Vertex,
+    /// 1-cell orbit.
+    Edge,
+    /// 2-cell orbit.
+    Face,
+    /// User-defined orbit. The integers correspond to the indices of the
+    /// beta functions that should be used for the computation.
+    Custom(&'a [u8]),
```
Comment (cedricchevalier19) [inline honeycomb-core/src/orbits.rs:27]:
```suggestion
    /// Ordered array of beta functions that define the orbit
    Custom(&'a [u8]),
```

Not sure it is mathematically correct, but I think the order thing is important.

**Review by cedricchevalier19 — APPROVED** [review]
Not totally sure about the naming.

File: honeycomb-core/src/orbits.rs (hunk 4 above)
Comment (imrn99) [reply honeycomb-core/src/orbits.rs]: This is actually a much better description!

File: CHANGELOG.md (hunk 1 above)
Comment (imrn99) [reply CHANGELOG.md:18]: I renamed `beta_bis` to `beta_runtime` and cahnged its implementation to use a match block. I don't wan't to rename the original beta function at the moment though.

File: honeycomb-core/src/twomap.rs (hunk 2 above)
Comment (imrn99) [reply honeycomb-core/src/twomap.rs:105]: I tried this initially, but it seems the compiler isn't fond of having a returned value in doc tests. I'm open to suggestion on how to resolve this

Comment (cedricchevalier19) [reply honeycomb-core/src/twomap.rs:105]: You can look at Métis-rs or coupe for examples.

**Pull request merged**